### PR TITLE
[SAP] Revisit backtracking line search

### PR DESCRIPTION
**Pull request by amcastro-tri**
A small cleanup before we add the exact line search.
This cleanup essentially avoids cases in which the gradient of the cost is so close to zero that small round-off errors trigger a false non-monotonic decrease of the cost.

We also implement the gradient of the cost for the line search, which is require for the exact line search.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408)

### CODE REVIEW

**Review by amcastro-tri — COMMENTED** [review]
+@xuchenhan-tri for feature review please

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408#-:-N4htB1J-iY0OU7KGmX-:bcllzq7)* status: LGTM missing from assignee xuchenhan-tri(platform), needs at least two assigned reviewers (waiting on @xuchenhan-tri)

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408)* status: 2 unresolved discussions, needs at least two assigned reviewers (waiting on @xuchenhan-tri)

___
*[`multibody/contact_solvers/sap/sap_solver.h` line 126 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4iOJLE3VNjgkHd8cd9:-N55WHaX66-9srJ09YYN:b1jp6tl) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.h#L126)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

It seems like this parameter only has an effect in debug mode? If so, we should label it that way.
</blockquote></details>

Also in release. When `nonmonotonic_convergence_is_error=true` the solver throws an exception if the convergence is not monotonic.
I added more docs on this.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 166 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4hxffS5qd8Rtglf_ms:-N55X0nIA5FPnabgA4cC:b-s6ys2q) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.cc#L166)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

According to the description of the flag in the header, this should "trigger an exception if the cost increases even if within the margins of
the slop." 

So the predicate should be a little different.

</blockquote></details>

True, I updated the docs. My intention for now is to let us play with this `slop` for a while until we are truly happy with its numerical values. In my stress test experiments these checks work great but I just wanted to make sure users do not get exceptions due to round-off errors when the solver could still make progress. 
That is why for now I kept `nonmonotonic_convergence_is_error=false` by default.

Long story short, I removed "even if within the margins of the slop.", which is wrong as per this PR.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 333 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4iPXYj9op8ueTAawZ6:-N55aQ8L3s5zflL4YA1-:b-2cw0db) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.cc#L333)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I'm not understanding why we return 1.0 in one case and alpha (which should assume the value of `alpha_max` here)  in another. Could you explain?
</blockquote></details>

Nope, I can't. I actually realize that the return with `1.0` is not a good idea. I only left the check on `dell` and I let the line search to try its best even if `dell_dalpha0` is close to zero.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 333 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4iNWULApw2VdhyFMS_:-N55au6h7du47ZyQrGiN:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.cc#L333)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

nit is abs needed here?
</blockquote></details>

Done

**Review by amcastro-tri — COMMENTED** [review]
+@jwnimmer-tri for platform review please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408#-:-N55rzfM4aZBd6UNfx-e:b-qxjhdt)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @jwnimmer-tri and @xuchenhan-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: platform.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408#-:-N55zY3X36RONGSb-qA2:b-vtj4vo)* status: 7 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[`multibody/contact_solvers/sap/sap_solver.h` line 128 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55xOu_DQ1AxxRqU2bs:-N55xOu_DQ1AxxRqU2bt:b-1t2i3x) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.h#L128)):*
> ```objc
>   // It is also used to check for monotonic convergence. In particular, we allow
>   // a small increase in the cost due to round-off errors
>   //   ℓᵏ < ℓᵏ⁻¹ + ε
> ```

nit From my reading of the implementation, I believe the satisfactory condition  is `<=` (not strictly `<`)?



_Code quote:_
```objc
  ℓᵏ < ℓᵏ⁻¹ + ε
```
___
*[`multibody/contact_solvers/sap/sap_solver.h` line 131 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55whaX9S70wzEx67sL:-N55whaX9S70wzEx67sM:bb7phvt) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.h#L131)):*
> ```objc
>   // where ε = relative_slop*max(1, (ℓᵏ+ℓᵏ⁻¹)/2).
>   // If this condition is not satified and nonmonotonic_convergence_is_error =
>   // true, SAPSolver throws an exception.
> ```

typo



_Suggestion:_
```objc
SapSolver
```
___
*[`multibody/contact_solvers/sap/sap_solver.h` line 135 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55wrRbCsELorrd48LL:-N55wrRc710SStFhO4FV:b-7ilvfu) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.h#L135)):*
> ```objc
> 
>   // (For debugging) Even though SAP's convergence in monotonic, round-off
>   // errors could cause small cost increases in the order of machine epsilon.
> ```

typo?



_Suggestion:_
```objc
on
```
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 136 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55yDU9EophyCikvGGP:-N55yDU9EophyCikvGGQ:b-1hn501) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.cc#L136)):*
> ```cpp
>   bool converged = false;
>   double alpha = 1.0;
>   int ls_iters = 0;
> ```

We are not reporting on the contents of a unix directory (`ls`).  After some digging some 60 lines later in this function, I discovered that we are doing a line search. 

https://drake.mit.edu/styleguide/cppguide.html#General_Naming_Rules



_Suggestion:_
```cpp
num_line_search_iters
```
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 162 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55ymvRBRyCzRgQ-_s0:-N55ymvS5vcOPXb7N7Kh:bb1e1lx) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.cc#L162)):*
> ```cpp
>       if (ell > ell_previous + ell_slop) {
>         DRAKE_LOGGER_DEBUG(
>             "At iter {} cost increased by: {}. alpha= {}. Relative momentum "
> ```

nit Is the inconsistent whitespace here on purpose?



_Suggestion:_
```cpp
alpha =
```

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408)* status: 6 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @amcastro-tri)

___
*[`multibody/contact_solvers/sap/sap_solver.h` line 130 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N565gJe0QSCcKh_TI57:-N565gJfEjrpDt2L15dG:b-400qea) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.h#L130)):*
> ```objc
>   //   ℓᵏ < ℓᵏ⁻¹ + ε
>   // where ε = relative_slop*max(1, (ℓᵏ+ℓᵏ⁻¹)/2).
>   // If this condition is not satified and nonmonotonic_convergence_is_error =
> ```

nit s/satified/satisfied.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 166 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4hxffS5qd8Rtglf_ms:-N566cOO8IQj7FEvd_28:bua4id3) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.cc#L166)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

True, I updated the docs. My intention for now is to let us play with this `slop` for a while until we are truly happy with its numerical values. In my stress test experiments these checks work great but I just wanted to make sure users do not get exceptions due to round-off errors when the solver could still make progress. 
That is why for now I kept `nonmonotonic_convergence_is_error=false` by default.

Long story short, I removed "even if within the margins of the slop.", which is wrong as per this PR.

</blockquote></details>

Gotcha, thanks for the explanation. In that case, I'd recommend putting the word "SAP" somewhere in the throw message (so that the user knows where the issue happened and who to ask for help from in case they get this message -- you).

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408)* status: 2 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @jwnimmer-tri)

___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 166 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N4hxffS5qd8Rtglf_ms:-N56eZLd3NrAZ3Ux-Gw7:bqtb2qt) ([raw file](https://github.com/robotlocomotion/drake/blob/396ea4b3bf708b009c24f4bb3a57d65fc661d6e4/multibody/contact_solvers/sap/sap_solver.cc#L166)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Gotcha, thanks for the explanation. In that case, I'd recommend putting the word "SAP" somewhere in the throw message (so that the user knows where the issue happened and who to ask for help from in case they get this message -- you).
</blockquote></details>

good idea. Done.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 136 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55yDU9EophyCikvGGP:-N56eFS74MbsRn8AwfKE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.cc#L136)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

We are not reporting on the contents of a unix directory (`ls`).  After some digging some 60 lines later in this function, I discovered that we are doing a line search. 

https://drake.mit.edu/styleguide/cppguide.html#General_Naming_Rules
</blockquote></details>

Done.
___
*[`multibody/contact_solvers/sap/sap_solver.cc` line 162 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17408#-N55ymvRBRyCzRgQ-_s0:-N56eKTB0nm-_TMYYRoN:b-jfyi89) ([raw file](https://github.com/robotlocomotion/drake/blob/544f0777ce829598f1351f633862b77243f78792/multibody/contact_solvers/sap/sap_solver.cc#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Is the inconsistent whitespace here on purpose?
</blockquote></details>

yes, I did it to upset the platform reviewer :-P

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17408)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @amcastro-tri)
